Firestore/core/src/firebase/firestore/nanopb/reader.h

```diff
@@ -134,6 +139,31 @@ class Reader {
   pb_istream_t stream_{};
 };
 
+class GrpcByteBufferReader {
```

Comment:
This class isn't particularly elegant, though it gets the job done. Will be happy to hear any suggestions. Some issues I have with it:
* the main one is where it should live. I don't see a clear argument for either gRPC layer or Nanopb layer, since it mixes both;
* it has to partially duplicate `Reader` (I somewhat mitigated this by using composition). The most straightforward way to avoid this is to make `ReadNanopbMessage` virtual. I'm a little hesitant since it would be a performance-unfriendly change, though.

Reply:
It's a single virtual function call so I wouldn't worry about it. We won't be able to measure the difference.

As far as layering goes, I think there's a dependency argument in favor of putting this in `remote`. `nanopb` is a pretty fundamental building block in the system so adding gRPC as a dependency here makes everything depend on gRPC.

As far as class structure goes: `Reader` wasn't made virtual initially because there wasn't any difference in the implementations (like there are for `Writer`). Making it parallel `Writer` in this regard seems like the way to go. In particular:
  * Make `Reader` a base class, leaving the error handling/context stuff in it
  * Make `StringReader`? for reading from strings/bytes
  * Make `ByteBufferReader` also derive from `Reader`.

In a follow up PR we can actually implement `pb_istream_t` in terms of the vector of slices, which would avoid the copy into `ByteString`.